let the `RuleOptimizer` optimize more rules

Palatis opened the pull request:
1. optimize `RecursiveRule` **probably needs review**
the idea is to return a new `RecursiveRule` and fetch the optimized version of the original `Rule` on-demand.
the original rule is optimized (if not yet) when fetching `CachedRule`, so we have to add the newly instantiated `RecursiveRule` to the LUT first, or we run into infinite recursion...
this might have some memory overhead, since we reference the LUT, the whole `RuleOptimizer` instance cannot be garbage-collected.
we can probably combat this by discarding `RecursiveRule.RuleFunc` after `RecursiveRule.CachedRule` is resolved, thus the referenced `RuleOptimizer` get's a chance to be GC-ed.

2. optimize the inner `Rule` of `NodeRule`
don't just stop when seeing a `NodeRule`, else its body don't get optimized.

3. optimize `OneOrMoreRule`
just like `ZeroOrMoreRule`.
   - A?+ => A* // rule.Optional().OneOrMore() => rule.ZeroOrMore()
   - A*? => A* // rule.ZeroOrMore().OneOrMore() => rule.ZeroOrMore()
   - A++ => A{2,} // rule.OneOrMore().OneOrMore() => rule.Counted(2, int.MaxValue)

4. optimize `CountedRule`
just like `OneOrMoreRule` and `ZeroOrMoreRule`
   - A{0,0} => _FALSE_ // rule.Counted(0, 0) => false
   - A{0,1} => A? // rule.Counted(0, 1) => rule.Optional()
   - A{0,} => A* // rule.Counted(0, int.MaxValue) => rule.ZeroOrMore()
   - A{1,) => A+ // rule.Counted(1, int.MaxValue) => rule.OneOrMore()

5. optimize `BooleanRule` **probably needs review**
   1. AtRule:
       - &_TRUE_ => _TRUE_ // rule.At(true) => true
       - &_FALSE_ => _FALSE_ // rule.At(false) => false
   2. NotAtRule:
       - !_TRUE_ => _FALSE_ // true.Not() => false
       - !_FALSE_ => _TRUE_ // false.Not() => true
   3. SequenceRule:
       - _TRUE_+A => A // true + rule => rule
       - _FALSE_+A => _FALSE_ // false + rule => false
       - A+_TRUE_ => A // seq + true => rule
       - A+_FALSE_ => _FALSE_ // rule + false => false
   4. ChoiceRule:
       - _TRUE_|A => _TRUE_ // true | rule => true
       - _FALSE_|A => A // false | rule => rule
       - A|_TRUE_ => _TRUE_ // rule | true => true
       - A|_FALSE_ => A // rule | false => rule

with these optimizations, `JsonTests.CompareToNewtonsoft()` yields ~10% speed-ups on my machine.
DEBUG version of the test went from ~16s to ~14s.
this is probably mostly because the `NodeRule.Body()` are getting optimized, now.

I'll post the tests in another PR.
```
File size = 32706k
It took 00:00:01.2276041 to parse using Newtonsoft
It took 00:00:16.5802442 to parse using Parakeet
Inner tree nodes = 3265209
It took 00:00:00.0073892 to optimize the rule
It took 00:00:14.6678079 to parse optimized using Parakeet
Inner tree nodes = 3265209
```

Palatis committed:
[optimizer] optimize inner rule of the `NodeRule`

don't just stop at `NodeRule`, also optimize its inner `Rule`.

Palatis committed:
[optimizer] optimize `OneOrMoreRule`

- A?+ => A* // rule.Optional().OneOrMore() => rule.ZeroOrMore()
- A*? => A* // rule.ZeroOrMore().OneOrMore() => rule.ZeroOrMore()
- A++ => A{2,} // rule.OneOrMore().OneOrMore() => rule.Counted(2, int.MaxValue)

Palatis committed:
[optimizer] optimize `BooleanRule`

1. AtRule:
    - &_TRUE_ => _TRUE_ // rule.At(true) => true
    - &_FALSE_ => _FALSE_ // rule.At(false) => false
2. NotAtRule:
    - !_TRUE_ => _FALSE_ // true.Not() => false
    - !_FALSE_ => _TRUE_ // false.Not() => true
3. SequenceRule:
    - _TRUE_+A => A // true + rule => rule
    - _FALSE_+A => _FALSE_ // false + rule => false
    - A+_TRUE_ => A // seq + true => rule
    - A+_FALSE_ => _FALSE_ // rule + false => false
4. ChoiceRule:
    - _TRUE_|A => _TRUE_ // true | rule => true
    - _FALSE_|A => A // false | rule => rule
    - A|_TRUE_ => _TRUE_ // rule | true => true
    - A|_FALSE_ => A // rule | false => rule

Palatis commented:
the force-push just fix commit logs...

Palatis commented:
i don't currently have an idea how to handle nested `CountedRule`, ie. `rule.Counted(3, 5).Counted(10, 20)`...

Palatis commented:
the `booleanRule.At()` and `BooleanRule.NotAt()` might not be that of a common case, but the `SequenceRule` and `ChoiceRule` merges actually do.

for example we can't do `rule + null`, so `WS => null;` will fail most of the time when we actually don't want WS.
thus we write`WS => BooleanRule.True;` instead, to make `SequenceRule` constructor happy.

however forexample `Node(rule)`, `Symbol("sym")`, or `Keyword("kw")` actually tries to constructs a sequence rule of `rule + WS`, then optimized into `rule + true`, which is bogus IMHO.
thus we optimize it further out to just `rule`.

Palatis commented:
the `booleanRule.At()` and `BooleanRule.NotAt()` might not be that of a common case, but the `SequenceRule` and `ChoiceRule` merges actually do.

for example we can't do `rule + null`, so `WS => null;` will fail most of the time when we actually don't want WS.
thus we write`WS => BooleanRule.True;` instead, to make `SequenceRule` constructor happy.

however forexample `Node(rule)`, `Symbol("sym")`, or `Keyword("kw")` actually tries to constructs a sequence rule of `rule + WS`, then optimized into `rule + true`, which is bogus IMHO.
thus we optimize it further out to just `rule`.

cdiggins reviewed (commented):
I appreciate your help and contributions. 

Unfortunately I don't have the bandwidth to properly review and test this set of changes right now. Optimizing the parser and maintaining correctness was quite challenging for me. Things now are relatively stable and I am concerned about the risk that we introduce new regressions. 

Is there a motivation for this work? Are things too slow? Or is this more of an intellectual curiosity?

Palatis commented:
of course I can craft the CST interfaces / classes by hand, however everytime i change something in the Grammar i also have to update the CST interfaces / classes by hand.
where i rather spend time writing a tool to generate them for me instead of modify the CST codes by myself.

Palatis commented:
I can break the PRs for individual rules, maybe that's easier to review.

cdiggins commented:
The SyntaxTreeGenerator sounds like a good idea, but I see nothing stopping you from creating it outside of the main project. I suggest that you go ahead and do build it and the new optimizer in a separate project that builds on top of the parts of Parakeet that you find useful, and when it is stable and working well, we can either: 
1. merge the changes back, or 
2. we can point people to your project a nice successor to parakeet.

As you might have noticed, I am not actively working on Parakeet, and since it is currently serving my needs I'm nervous about making too many changes to it.

cdiggins commented:
> I can break the PRs for individual rules, maybe that's easier to review.

My problem is that there is no promise it is safe to merge the optimizations. You just "think it might be". I'd like to see it used more extensively first. I wish I had a better testing system in place, but I don't.  

I'd also like to reduce the churn because I have a lot of other projects that need my attention right now.

Palatis commented:
> The SyntaxTreeGenerator sounds like a good idea, but I see nothing stopping you from creating it outside of the main project. I suggest that you go ahead and do build it and the new optimizer in a separate project that builds on top of the parts of Parakeet that you find useful, and when it is stable and working well, we can either:
> 
>     1. merge the changes back, or
>     2. we can point people to your project a nice successor to parakeet.
> 
> As you might have noticed, I am not actively working on Parakeet, and since it is currently serving my needs I'm nervous about making too many changes to it.

sure, I'm doing this in the hope that somebody else would find these useful, and to make the world better.
it's absolutely okay if you would accept or reject any of the PRs i submitted.

i understand that optimizing rules is a big topic and requires lots of code reviews.
it's totally okay if you don't feel comfortable merging any of these.

Palatis committed:
[optimizer] optimize `CountedRule`

- A{0,1} => A? // rule.Counted(0, 1) => rule.Optional()
- A{0,} => A* // rule.Counted(0, int.MaxValue) => rule.ZeroOrMore()
- A{1,1} => A // rule.Counted(1, 1) => rule
- A{1,) => A+ // rule.Counted(1, int.MaxValue) => rule.OneOrMore()

Palatis committed:
Merge branch 'rfe/optimizers/BooleanRule' into rfe/optimizer

Palatis committed:
Merge branch 'rfe/optimizers/OneOrMoreRule' into rfe/optimizer

Palatis committed:
Merge branch 'rfe/optimizers/CountedRule' into rfe/optimizer

Palatis committed:
Merge branch 'rfe/optimizers/NodeRule' into rfe/optimizer
